### PR TITLE
Base Discrete Maths Package

### PR DESCRIPTION
**Pull request by Retr05041**
Created a discrete maths cog and discrete maths package, along with a simple logical operations module for said package which includes:

- [x] Gates for: negation, conjunction, disjunction, implies, bi-implication, xor, nand, nor, xnor
- [x] Truth Table generator
- [x] Tautology checker

### CODE REVIEW

File: lib/discrete_maths/logic_operations.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        variable_values = {}
+        for i in range(len(variable_list)):
+            variable_values[variable_list[i]] = combination[i]
+        results.append(calculate_rpn_prop(rpn_tokenization, variable_values))
+
+    # Check if all the results are true
+    return all(results)
+
+# Create a truth table from a proposition
+def create_truth_table(proposition: str) -> str:
```
Comment (PenguinBoi12) [inline lib/discrete_maths/logic_operations.py:137]: The logic that creates the table shouldn't be here. It should be in the cog. However, the logic that evaluates the logical proposition can/should stay here. 

The reason why is that this library shouldn't care too much about how the evaluation of a proposition is used. It should only care about evaluating it and returning some results. Who cares about how it's used and what is displayed is the cog. The cog is the link between the logic and the user. 

In other words, all the sanity checkers and logic for the evaluation is fine here but the code to draw (this includes making the variables) the table shouldn't be here but in the cog.

File: lib/discrete_maths/logic_operations.py
Comment (PenguinBoi12) [inline lib/discrete_maths/logic_operations.py:1]: I'd change the name for something else that represent a bit more clearly what it does (which is to evaluate logical proposition). Right now, `logical_operation` isn't very descriptive. 

Also, it's in it's own package but it's alone? Wouldn't the parser module be included? Since this parser is built for this exactly that would be interesting to include it. Imagine if someone says "I want to take this libraries to use it somewhere else" but the parser is missing. It would be a bit awkward. So either move the parser there since it's for that only or move it somewhere else.

File: lib/discrete_maths/logic_operations.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
+# Classical logic operations library
+
+from lib.parsers import shunting_yard
+
+# Operator Precedence Dictionary - = == iff, > == implies, v == or, ^ == and, ~ == not
+OP_PRECIDENCE = {"=": 2, ">": 3, "v": 4, "^": 5, "~": 6}
```
Comment (PenguinBoi12) [inline lib/discrete_maths/logic_operations.py:6]: There's a typo here. It's `precedence`. Additionally it would be better if it was called `OPERATION_PRECEDENCES` or `OPERATION_BY_PRECEDENCES` (the use of `by` is commonly used in hashes to describe what it is)

File: lib/discrete_maths/logic_operations.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+            stack.append(disjunction(stack.pop(), stack.pop()))
+        elif token == ">":
+            stack.append(implies(stack.pop(), stack.pop()))
+        elif token == "=":
+            stack.append(iff(stack.pop(), stack.pop()))
+    
+    return stack.pop()
+
+# Create a list of lists of all the possible values for the variables in the rpn proposition
+def create_variable_values(rpn: list) -> list:
```
Comment (PenguinBoi12) [inline lib/discrete_maths/logic_operations.py:97]: I think `create_values_for_variables` or `generate_values_for_variables` would be more meaningful and would allow you to not have to comment it and explain it.  

This is a function to create all combination of values? You can simplify this a lot. For example you can use regex to extract all variables. Also, you can do part of the process directly while creating the truth table instead of doing the process of creating a whole list of combination. It's much easier and you're doing everything at the same time instead of doing it multiple time. 

Also, see comment for `create_truth_table`. This function is part of it.

**Review by PenguinBoi12 — CHANGES_REQUESTED** [review]
I'm going to start by talking about the comments. Comments provide additional information that the code itself cannot convey. They should add context to what you're doing or explain complex parts of the code when necessary. Consequently, it's important to avoid redundant comments that merely restate what is already evident from the code's readability. Comments should be reserved for situations where the code is not self-explanatory. 

Here's some example of a useless comments.
```python
# add a and b
def add(a, b):
    return a + b

# imports
from ... import ...
from ... import ...
import ....

# defines my_function
def my_function():
   ...
```

On another note, in Python you want to use [doscstring](https://www.programiz.com/python-programming/docstrings) when commenting functions, methods, classes or modules.


The second recurring issue I see is the use of abbreviation. Stuff like `rpn`, `op`, `prop`, etc. It's a bad habit. Abbreviation should be avoided as much as possible; there's only a few cases where they are accepted. Change them for full words.

As for the feature itself, it seems to work well. However, it would be nice to have a command to see how to use it. A sort of guide; we don't really know which symbols to use without looking at the code.

File: lib/discrete_maths/logic_operations.py
Comment (PenguinBoi12) [reply lib/discrete_maths/logic_operations.py:1]: This is a comment
